umlfri2/application/dispatcher.py

```diff
@@ -1,34 +1,42 @@
+from __future__ import annotations
+
 from _weakref import ref
 from types import MethodType
+from typing import Any, Callable, Dict, List, Optional, Type, TYPE_CHECKING, Union
+
+if TYPE_CHECKING:
+    from umlfri2.application import Application
+    from umlfri2.application.events.base import Event
 
 
 class methodref(ref):
-    def __new__(cls, method, callback=None):
+    def __new__(cls, method: MethodType, callback: Optional[Callable[[ref], Any]] = None) -> methodref:
         self = ref.__new__(cls, method.__self__, callback)
         self.__fnc = method.__func__
         return self
     
-    def __call__(self):
+    def __call__(self) -> Optional[MethodType]:
         obj = super().__call__()
         if obj is None:
             return None
         return MethodType(self.__fnc, obj)
 
 
 class EventDispatcher:
-    def __init__(self, application):
+    def __init__(self, application: Application) -> None:
         self.__application = application
-        self.__events = {}
+        self.__events: Dict[Optional[Type[Event]], List[Union[ref, Callable[[Event], None]]]] = {}
     
-    def subscribe(self, event_type, function, auto_unsubscribe=True):
+    def subscribe(self, event_type: Optional[Type[Event]], function: Callable[[Event], None],
```

Comment:
subscribe/unsubscribe could be made generic, as event_type represent the type of the only argument of the function callable